internal/cmd/login.go

```diff
@@ -154,116 +156,57 @@ func login(cli *CLI, options loginCmdOptions) error {
-		}
+		loginRes, err = lc.APIClient.Login(ctx, &api.LoginRequest{
+			PasswordCredentials: &api.LoginRequestPasswordCredentials{
+				Name:     options.User,
+				Password: options.Password,
+			},
+		})
+		if err != nil {
+			if api.ErrorStatusCode(err) == http.StatusUnauthorized || api.ErrorStatusCode(err) == http.StatusNotFound {
+				return &LoginError{Message: "your username or password may be invalid"}
+			}
 
+			return err
+		}
 	default:
 		if options.NonInteractive {
 			return Error{Message: "Non-interactive login requires setting either the INFRA_ACCESS_KEY or both the INFRA_USER and INFRA_PASSWORD environment variables"}
 		}
 
-		if err = checkDeviceFlowCompatibility(ctx, lc.APIClient); err != nil {
-			return err
-		}
-
-		resp, err := deviceFlowLogin(ctx, lc.APIClient, cli)
-		if err != nil {
-			return err
-		}
-
-		loginReq.AccessKey = resp.AccessKey
-		err = updateInfraConfig(lc, resp)
+		loginRes, err = deviceFlowLogin(ctx, lc.APIClient, cli)
 		if err != nil {
 			return err
 		}
 	}
 
-	if err := loginToInfra(cli, lc, loginReq, options.NoAgent); err != nil {
-		return err
-	}
-	if options.InjectUserSSHConfig {
-		return updateUserSSHConfig(cli)
-	}
-	return nil
-}
-
-func checkDeviceFlowCompatibility(ctx context.Context, api *api.Client) error {
-	version, err := api.GetServerVersion(ctx)
-	if err != nil {
-		return err
-	}
-
-	// append -0 to compare against "prerelease" versions
-	// see https://pkg.go.dev/github.com/Masterminds/semver/v3#hdr-Checking_Version_Constraints_and_Comparing_Versions
-	c, err := semver.NewConstraint(fmt.Sprintf(">= %s-0", DeviceFlowMinVersion))
-	if err != nil {
-		return err
-	}
-
-	v, err := semver.NewVersion(version.Version)
-	if err != nil {
-		return err
-	}
-
-	if !c.Check(v) {
-		return Error{
-			Message: fmt.Sprintf("Your version of Infra Server (%s) is out of date. Please upgrade to %s or later.", v, DeviceFlowMinVersion),
-		}
-	}
-
-	return nil
-}
-
-func equalHosts(x, y string) bool {
-	if x == y {
-		return true
-	}
-	if strings.TrimPrefix(x, "https://") == strings.TrimPrefix(y, "https://") {
-		return true
-	}
-	return false
-}
-
-func loginToInfra(cli *CLI, lc loginClient, loginReq *api.LoginRequest, noAgent bool) error {
-	ctx := context.TODO()
-	loginRes, err := lc.APIClient.Login(ctx, loginReq)
-	if err != nil {
-		if api.ErrorStatusCode(err) == http.StatusUnauthorized || api.ErrorStatusCode(err) == http.StatusNotFound {
-			switch {
-			case loginReq.PasswordCredentials != nil:
-				return &LoginError{Message: "your username or password may be invalid"}
-			}
-		}
-
-		return err
-	}
 	// Update the API client with the new access key from login
 	lc.APIClient.AccessKey = loginRes.AccessKey
 
 	if loginRes.PasswordUpdateRequired {
 		fmt.Fprintf(cli.Stderr, "  Your password has expired. Please update your password.\n")
 
-	PROMPTLOGIN:
-		password, err := promptSetPassword(cli, loginReq.PasswordCredentials.Password)
-		if err != nil {
-			return err
-		}
+		for {
```

Comment:
I _think_ this is fine. It's definitely better w/ the `for` loop here than the `goto`. I'm trying to think if we would ever break out of here on too many attempts (I mean, the user can always CTRL-C, so we have to handle the case anyway).